stdpopsim/annotations.py

```diff
@@ -0,0 +1,165 @@
+
+    def __str__(self):
+        s = "GTF Annotation:\n"
+        s += "\tspecies   = {}\n".format(self.species.name)
+        s += "\tid        = {}\n".format(self.id)
+        s += "\turl       = {}\n".format(self.url)
+        s += "\tzarr url       = {}\n".format(self.zarr_url)
+        s += "\tcached    = {}\n".format(self.is_cached())
+        s += "\tcache_dir = {}\n".format(self.species_cache_dir)
+        return s
+
+    def is_cached(self):
+        """
+        Returns True if this map is cached locally.
+        """
+        return os.path.exists(self.species_cache_dir)
+
+    def download(self):
+        """
+        Downloads the zarr from the source URL and stores it in the
+        cache directory. If the annotation directory already exists it is first
+        removed.
+        """
+        self.file_name = os.path.basename(self.zarr_url)
+        if self.is_cached():
+            logger.info(f"Clearing cache {self.species_cache_dir}")
+            for fn in os.listdir(self.species_cache_dir):
+                file_path = os.path.join(self.species_cache_dir, fn)
+                try:
+                    os.unlink(file_path)
+                except OSError:
+                    warnings.warn(f'could not unlink {file_path}')
+        logger.debug(f"Checking species cache directory {self.species_cache_dir}")
+        os.makedirs(self.species_cache_dir, exist_ok=True)
+        download_file = f'{self.species_cache_dir}/{self.file_name}'
+        print(download_file)
+        logger.info(f"Downloading Zarr file '{self.id}' from {self.zarr_url}")
+        if os.path.exists(download_file):
+            warnings.warn("multiple downloads?")
+        try:
+            x, y = urllib.request.urlretrieve(self.zarr_url, download_file)
+        except FileNotFoundError:
+            warnings.warn("can't connnect to url")
+
+        logger.debug("Download Zarr complete")
+        logger.info(f"Storing Zarr in {self.species_cache_dir}")
+
+    def get_chromosome_annotations(self, id):
+        """
+        Returns the pandas dataframe for
+        the chromosome with the specified id.
+        """
+        if not self.is_cached():
+            self.download()
+        annot_file = os.path.join(self.species_cache_dir, self.file_name)
+        if id is None:
+            raise ValueError("bad chrom id")
+        chr_prefix = "chr"  # building this in for future generalization
+        if id.startswith(chr_prefix):
+            id = id[len(chr_prefix):]
+        if os.path.exists(annot_file):
+            bed = zarr_to_dataframe(annot_file)
+            assert type(bed) == pandas.DataFrame
+            ret = bed[bed.seqid == id]
+            if len(ret) == 0:
+                raise ValueError
+        else:
+            warnings.warn(
+                "Warning: annotation file not found for chromosome: '{}'"
+                " on annotation: '{}', no annotations will be used".format(
+                    id, self.id))
+            ret = None
```

Comment:
It isn't obvious to me how this case can happen. But I think it would be better if this raised a ValueError, instead of having a warning and returning None.